### DISCUSSION TITLE
Missing annotations

### DISCUSSION THREAD

**Discussion by favonia** [artifacthub/hub#2945] [Q&A] [answered]
I have received an email from Artifact Hub indicating that there is an error with my published Docker images. The error message reads as follows:
```
error getting repository metadata: failed to authorize: failed to fetch oauth token: Post "https://auth.docker.io/token": EOF
```
I used the OCI address `oci://index.docker.io/favonia/cloudflare-ddns` and got the above message. I also attempted using `registry-1.docker.io`, but it seemed to lead to the same error. I am uncertain about how to resolve this issue. Could someone please advise me on what steps to take?

**Comment by tegioz** [comment, marked as answer]
Hi @favonia 👋 

It looks like AH found some errors while processing your package (please see below). You should be able to see those errors from the control panel in the UI (and receive an email notification about them). There is some required metadata your image needs to include before being able to list it on AH. For more details please see: <https://artifacthub.io/docs/topics/repositories/container-images/>

```
error preparing package (tag: edge): 3 errors occurred:                            
     * required metadata field not provided: org.opencontainers.image.created  
     * required metadata field not provided: org.opencontainers.image.description
     * required metadata field not provided: io.artifacthub.package.readme-url 

error preparing package (tag: latest): 3 errors occurred:                           
     * required metadata field not provided: org.opencontainers.image.created  
     * required metadata field not provided: org.opencontainers.image.description
     * required metadata field not provided: io.artifacthub.package.readme-url 
```

**Reply by favonia** [reply]
@tegioz Thank you for taking a look, but my images do have all the labels. (The errors happened only recently, and the CI script has not been changed.) The real error is that the AH failed to fetch an OAuth token from Docker Hub, and thus it could not fetch _any_ label. The email I received has a more complete log; in particular, the first line (which indicates the real error) was _not_ publicly shown:
```
error getting repository metadata: failed to authorize: failed to fetch oauth token: Post "https://auth.docker.io/token": EOF

error preparing package (tag: edge): 3 errors occurred:
     * required metadata field not provided: org.opencontainers.image.created
     * required metadata field not provided: org.opencontainers.image.description
     * required metadata field not provided: io.artifacthub.package.readme-url

error preparing package (tag: latest): 3 errors occurred:
     * required metadata field not provided: org.opencontainers.image.created
     * required metadata field not provided: org.opencontainers.image.description
     * required metadata field not provided: io.artifacthub.package.readme-url
```

**Reply by tegioz** [reply]
No worries!

I've just added your container image package locally and the problem is the same. The oauth error could have been a punctual issue with Docker Hub, but it's not happening anymore, neither in artifacthub.io (where I got the previous logs from) or locally (please see screenshots below):

<img width="1606" alt="logs1" src="https://user-images.githubusercontent.com/1213902/230053499-9e6e6a5c-b27f-4162-81fc-c1422bf9db81.png">

<img width="1529" alt="logs2" src="https://user-images.githubusercontent.com/1213902/230053522-faf8744e-3090-4cba-9182-79fadde1ae89.png">

Have you seen your package listed on AH before? Let's try to find out why it's not picking that metadata then 🙂

**Reply by tegioz** [reply]
All right, I think I know what's going on 🙂

Your images seem to be using the OCI manifest format, so they should use annotations instead of labels. You can find more details about this in the [Containers images repositories guide](https://artifacthub.io/docs/topics/repositories/container-images/#image-metadata).

**Reply by favonia** [reply]
@tegioz Thank you. I was aware of labels v.s. annotations, but I was confused because the AH was able to pick up those labels until around ~~the end of the last year~~ the start of this year. I thought the GitHub Actions workflows I used managed to translate labels to annotations when pushing manifests. Apparently, either the AH had been tolerating, or the Docker Hub was changed, or the sub-workflows (maintained by others) have been changed. Is there any recent change about labels/annotations? I will re-check the workflows I am using now. Thank you again for your help.

**Reply by favonia** [reply]
The Actions I used, `docker/build-push-action`, does not seem to provide an easy way to add annotations other than some hack. (See https://github.com/docker/build-push-action/issues/447.) Thank you again for your time---I will find some way to add those annotations (though I still wonder why the AH was okay before, because apparently annotations were never added).

**Reply by tegioz** [reply]
No worries, my pleasure :)

Mmm the [code that handles this hasn't changed in the last couple of years](https://github.com/artifacthub/hub/blame/61701344a7478190fdbece705e19a44d774d0d33/internal/tracker/source/container/container.go#L379-L396). Is it possible that something has changed in the process of building your images (maybe an upgrade in any GitHub action)?

I've taken a look at your images, and you can see when the manifest format changed (from `1.8.2` to `1.8.3`), just in case that helps:

```
crane manifest favonia/cloudflare-ddns:1.8.2 | jq
{
  "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
```

```
crane manifest favonia/cloudflare-ddns:1.8.3 | jq
{
  "mediaType": "application/vnd.oci.image.index.v1+json",
```

The tag `1.8.2` works fine in AH actually, I've just tried it locally:

<img width="1563" alt="Screenshot 2023-04-05 at 14 50 49" src="https://user-images.githubusercontent.com/1213902/230085595-89cfc221-b563-4824-942d-bb97db6cea0e.png">

Hope that helps!

**Reply by tegioz** [reply]
> (though I still wonder why the AH was okay before, because apparently annotations were never added).

Because when you were using the previous manifest format (up to version `1.8.2`), the labels were the way to go 😉

**Reply by favonia** [reply]
@tegioz Thanks. So I guess `docker/build-push-action` changed the output manifest format recently. This is very helpful information.

**Reply by favonia** [reply]
@tegioz FYI: I created the issue https://github.com/docker/build-push-action/issues/854 based on what you have discovered.

**Reply by favonia** [reply]
~~For people who may have found this discussion, comprehensive support of annotations has since arrived in the upstream tools (including GitHub Actions) and it is now trivial to satisfy ArtifactHub. 😀~~ Actually, I take it back. There's still one final piece to make the support complete: https://github.com/docker/metadata-action/issues/360 After this I think it will be trivial to satisfy ArtifactHub.

**Reply by favonia** [reply]
@tegioz I wonder if I could have your help again. I still got the same error, but I believe I have added the annotations this time. You can see them by running the following command:
```bash
crane manifest favonia/cloudflare-ddns:edge | jq .annotations
```
Its output is:
```json
{
  "io.artifacthub.package.license": "Apache-2.0",
  "io.artifacthub.package.maintainers": "[{\"name\":\"favonia\",\"email\":\"favonia@email.com\"}]",
  "io.artifacthub.package.readme-url": "https://raw.githubusercontent.com/favonia/cloudflare-ddns/main/build/SUMMARY.markdown",
  "org.opencontainers.image.created": "2023-12-01T14:02:32.121Z",
  "org.opencontainers.image.description": "🌟 A small, feature-rich, and robust Cloudflare DDNS updater",
  "org.opencontainers.image.licenses": "Apache-2.0",
  "org.opencontainers.image.revision": "fe2ed0037ebba39bb7f2a4c594f58d462439a76f",
  "org.opencontainers.image.source": "https://github.com/favonia/cloudflare-ddns",
  "org.opencontainers.image.title": "cloudflare-ddns",
  "org.opencontainers.image.url": "https://github.com/favonia/cloudflare-ddns",
  "org.opencontainers.image.version": "edge"
}
```
I wonder if there's something obvious I missed? Thank you again for your help so far.

**Reply by favonia** [reply]
I might have figured it out... ArtifactHub does not seem to be reading annotations from the index manifest but some image manifest? Because the code is calling https://pkg.go.dev/github.com/google/go-containerregistry/pkg/v1/remote#Descriptor.Image and its documentation says:

> If the fetched artifact is an index, it will attempt to resolve the index to a child image with the appropriate platform. 

I can certainly add the same annotations at both levels for compatibility (EDIT: I did now), but I want to point out that this is different from what GitHub is doing, in case this is an intentional decision: https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry#adding-a-description-to-multi-arch-images

**Reply by tegioz** [reply]
> There's still one final piece to make the support complete: https://github.com/docker/metadata-action/issues/360 After this I think it will be trivial to satisfy ArtifactHub.

Thanks @favonia! Please feel free to submit a PR to update the [containers images repositories section in the repositories guide](https://artifacthub.io/docs/topics/repositories/container-images/) ([source](https://github.com/artifacthub/hub/blob/master/docs/container_images_repositories.md)) with any information that may be helpful to users, we'd really appreciate it 😄

> I can certainly add the same annotations at both levels for compatibility (EDIT: I did now), but I want to point out that this is different from what GitHub is doing, in case this is an intentional decision.

We'll take a look at this to see if we can improve it 🙂 Thanks!

**Reply by favonia** [reply]
@tegioz Thanks. Perhaps ArtifactHub has to decide what to do first. I guess you could either (1) keep the current status or (2) follow GitHub or (3) check the index manifest, and if annotations are missing there, check some image manifest. Each decision will lead to its own instructions. The thing is I haven't checked what other websites or software (Amazon/Azure/GitLab/Harbor/Quay/...) do. For that reason I don't want to claim what GitHub does is the best. This perhaps requires some internal study/discussion within ArtifactHub to decide what to do with these manifests in general.

**Reply by tegioz** [reply]
Sure, I meant just some minor notes based on your comments in this discussion and your experience during the process. We probably won't change how it works in the short term, so it'd be based on how it currently works. My point was that any update to that document that can make users life easier would be welcome 🙂

Adding support for containers images to AH was kind of an experiment, so please don't take our way as the recommended or best way either 😇

**Reply by favonia** [reply]
@tegioz The more I think about it, the more I feel ArtifactHub's current approach is a bit problematic. The reason is that it is fetching the annotations of the image matching the "default platform," which is `amd64/linux`, and if all children of an index explicitly label them as non-linux or non-amd64, ArtifactHub will err. I feel it's a bit strange if a registry only accepts a multi-arch build with amd64/linux, and I'm not sure how to document "your multi-arch container should have a build for amd64/linux". Could you possibly reconsider changing the current behavior? Should I open a GitHub issue/PR?

References for relevant code:
- https://github.com/google/go-containerregistry/blob/55ffb0092afd1313edad861a553b4fcea21b4da2/pkg/v1/remote/options.go#L129-L138
- https://github.com/google/go-containerregistry/blob/55ffb0092afd1313edad861a553b4fcea21b4da2/pkg/v1/remote/options.go#L59-L62
- https://github.com/google/go-containerregistry/blob/55ffb0092afd1313edad861a553b4fcea21b4da2/pkg/v1/remote/index.go#L173-L190

**Reply by tegioz** [reply]
Yes, we can reconsider changing how this works. I need to finish a couple of things first but I'll try to take a look at it soon and get back to you :)

**Reply by tegioz** [reply]
Hi @favonia 👋

AH now supports annotations in the index manifest and the default platform (`linux/amd64`) is not required anymore in multi-arch images 🙂

From <https://artifacthub.io/docs/topics/repositories/container-images/>:

<img width="940" alt="metadata" src="https://github.com/artifacthub/hub/assets/1213902/80eb6886-9cd9-414d-b625-081306c2aab9">

**Reply by favonia** [reply]
@tegioz Thank you! I can confirm that it is working for me.
